src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -584,22 +584,40 @@ public override bool Equals(object obj)
 
                 var other = obj as ReturnInferenceCacheKey;
 
-                return (object)other != null &&
-                    other.ParameterTypes.SequenceEqual(this.ParameterTypes) &&
-                    other.ParameterRefKinds.SequenceEqual(this.ParameterRefKinds) &&
-                    other.TaskLikeReturnTypeOpt == this.TaskLikeReturnTypeOpt;
+                if ((object)other == null ||
+                    other.ParameterTypes.Length != this.ParameterTypes.Length ||
+                    other.TaskLikeReturnTypeOpt != this.TaskLikeReturnTypeOpt)
+                {
+                    return false;
+                }
+
+                for (int i = 0; i < this.ParameterTypes.Length; i++)
+                {
+                    if (!other.ParameterTypes[i].Equals(this.ParameterTypes[i], TypeCompareKind.CompareNullableModifiersForReferenceTypes) ||
+                        other.ParameterRefKinds[i] != this.ParameterRefKinds[i])
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
             }
 
             public override int GetHashCode()
             {
-                return Hash.Combine(TaskLikeReturnTypeOpt?.GetHashCode() ?? 0, Hash.CombineValues(this.ParameterTypes));
+                var value = TaskLikeReturnTypeOpt?.GetHashCode() ?? 0;
+                foreach (var type in ParameterTypes)
+                {
+                    value = Hash.Combine(type.TypeSymbol, value);
```

Comment:
Arguments are `T newKeyPart, int currentKey`. The second argument is expected to be a hash code.